instrumentation/log4j/log4j-2.13.2/auto/src/main/java/io/opentelemetry/instrumentation/auto/log4j/v2_13_2/Log4j2MdcInstrumentation.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.auto.log4j.v2_13_2;
+
+import static io.opentelemetry.javaagent.tooling.ClassLoaderMatcher.hasClassesNamed;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+import com.google.auto.service.AutoService;
+import io.opentelemetry.javaagent.tooling.Instrumenter;
+import java.util.Collections;
+import java.util.Map;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+@AutoService(Instrumenter.class)
+public final class Log4j2MdcInstrumentation extends Instrumenter.Default {
+  public Log4j2MdcInstrumentation() {
+    super("log4j2", "log4j", "log4j-2.13.2");
+  }
+
+  @Override
+  public ElementMatcher<ClassLoader> classLoaderMatcher() {
+    return hasClassesNamed("org.apache.logging.log4j.core.util.ContextDataProvider");
+  }
+
+  @Override
+  public ElementMatcher<? super TypeDescription> typeMatcher() {
+    return named("org.apache.logging.log4j.core.impl.ThreadContextDataInjector");
```

Comment:
Change to `org.apache.logging.log4j.core.util.ContextDataProvider` for consistency?

Reply:
Unfortunately the instrumentation does not work if I use `ContextDataProvider` here: I suspect that it's because it's an interface.

Reply:
Then at least add the comment

Reply:
Interfaces are ok.

The problem with using `org.apache.logging.log4j.core.util.ContextDataProvider` is that then helper _class_ injection is performed during the loading (and transformation) of `ContextDataProvider`, which is a problem because one of the classes that we inject implements this same interface, causing the interface to be loaded while it's being transformed, which leads to duplicate class definition error after the interface is transformed and the triggering class loader tries to load it.

It seems this will be a common problem with injecting SPI resources, which is too bad, we'll need to be careful to pick type for the `typeMatcher` in these cases that we know will get loaded before SPI is triggered

Reply:
😮 Thanks for the explanation, it makes much more sense now!

I've replaced my comment with your explanation.